pythonforandroid/graph.py

```diff
@@ -99,17 +129,125 @@ def find_order(graph):
                 bset.discard(result)
 
 
-def get_recipe_order_and_bootstrap(ctx, names, bs=None):
-    recipes_to_load = set(names)
+def obvious_conflict_checker(ctx, name_tuples, blacklist=[]):
+    """ This is a pre-flight check function that will completely ignore
+        recipe order or choosing an actual value in any of the multiple
+        choice tuples/dependencies, and just do a very basic obvious
+        conflict check.
+    """
+    deps_were_added_by = dict()
+    deps = set()
+
+    # Add dependencies for all recipes:
+    to_be_added = [(name_tuple, None) for name_tuple in name_tuples]
+    while len(to_be_added) > 0:
```

Comment:
Minor: that loop is insanely large AND deep :cry: 
That function looks pretty complex in terms of cyclomatic complexity for "just do a very basic obvious conflict check"

Reply:
@AndreMiras *shrug* I don't think that is a surprise or even a problem. It basically builds the graph in a stupid manner, so it has the same runtime complexity (obviously) as the regular graph. However, it does it in a much simpler way in terms of how it deals with multiple possibilities, which is why it is a much nicer place for error handling

Think about the runtime in a maths fashion: it's pretty much "just" double. It's not exponentially larger or anything, and there is definitely no obvious way to get a better result unless you want to do something rather different than the actual graph code itself.

So I'm not sure what your comment intends to achieve. What would you rather have me do instead?

Reply:
I don't believe there is any reasonable way to reduce it. Or can you suggest one? 

As for breaking it down: splitting it up into multiple functions without actually reusing the code elsewhere IMHO  would actually **reduce** readability of the overall algorithm/code flow, with a function of this size which I actually still find straightforward enough - it's pretty much just a queue processing loop with relatively granular, commented steps inside, it's not that hard to understand or that complex in nesting. Littering that code into multiple functions wouldn't make it any simpler to read, again, IMHO. So I think that is a bad idea, unless you can show me a split up that actually looks decent & better to read, but I have a hard time imagining any

So the only other way to reduce it would be to remove functionality, unless there is some miracle approach I have missed. Or of course it could be removed and we could go back to dumb error messages.

What exactly did you have in mind?

Reply:
Just to be clearer/more concrete: I think putting 4-line code chunks (or even 5-6 lines) into a separate function is in general a bad idea unless it is actually reused somewhere. It creates more friction scrolling to that function and figuring out what it does in detail than you gain in readability, it's IMHO better to just put a comment over the chunk that says what it does (which I already did)

Reply:
http://number-none.com/blow/john_carmack_on_inlined_code.html CTRL+F for "I have historically used" to jump to the section where the benefits of not splitting up at least super tiny things are explained, if you're curious about some smarter writing than what I could ever argue on this topic

Reply:
Again I'm suggesting, as I said `break it down per concern when possible`. So it means what it means, if you think it's OK as it is then fine

Reply:
@AndreMiras hm alright. maybe I misinterpreted your first remark, sorry :smile: I can't think of a much better way to write it, but you know, I'm very biased since I wrote it :joy:  so if you happen to come up with a more concrete change, feel free to let me know

Reply:
Sure :+1: